src/compiler/checker.ts

```diff
@@ -24062,12 +24098,61 @@ namespace ts {
+                if (isIntersectionType(type)) {
+                    const objectSubtype = type.types.find(t => getObjectFlags(t) & ObjectFlags.Anonymous) as ObjectType | undefined;
+                    if (objectSubtype) {
+                        const restOfIntersection = filterUnionOrIntersectionType(type, t => t !== objectSubtype);
+                        return createIntersectionType([restOfIntersection, widenObjectType(objectSubtype, newSymbol)]);
+                    }
+                }
+
+                // Otherwise, just add the new object type as an intersection
+                const newTypeWithSymbol = widenObjectType(createAnonymousType(undefined, createSymbolTable(), emptyArray, emptyArray, emptyArray), newSymbol);
+                return createIntersectionType([type, newTypeWithSymbol]);
+
+                function widenObjectType(type: ObjectType, newSymbol: Symbol): Type {
+                    const members = createSymbolTable();
+                    if (type.members !== undefined) {
+                        mergeSymbolTable(members, type.members);
+                    }
+                    members.set(newSymbol.escapedName, newSymbol);
+                    return createAnonymousType(undefined, members, type.callSignatures ?? emptyArray, type.constructSignatures ?? emptyArray, type.indexInfos ?? emptyArray);
+                }
+            }
+
+            function narrowOrWidenTypeByInKeyword(type: Type, name: __String, assumeTrue: boolean) {
+                // If type contains global this, don't touch it
+                if (type.symbol === globalThisSymbol
+                    || isUnionOrIntersectionType(type) && filterUnionOrIntersectionType(type, t => t.symbol === globalThisSymbol) !== neverType
+                   ) {
+                    return type;
+                }
+                const someDirectSubtypeContainsProp = getPropertyOfType(type, name, /* skipObjectFunctionPropertyAugment */ false, /* includePartialProperties */ true);
+                if (someDirectSubtypeContainsProp) {
+                    // If union, filter out all components not containing the property
+                    // Otherwise, either return the type or never
+                    if (type.flags & (TypeFlags.Object | TypeFlags.UnionOrIntersection)
+                        || isThisTypeParameter(type)
+                    ) {
+                        return filterType(type, t => isTypePresencePossible(t, name, assumeTrue));
+                        return isTypePresencePossible(type, name, assumeTrue) ? type : neverType;
+                    }
+                }
+                // only widen property when the type does not contain string-index/name in any of the constituents.
+                if (assumeTrue && !someDirectSubtypeContainsProp && !getIndexInfoOfType(type, stringType)) {
+                    const addSymbol = createSymbol(SymbolFlags.Property, name);

```

Comment:
We should cache these symbols (and the resulting properties and single-property-types) globally within the checker. Object types aren't structurally cached, so it needs to be done explicitly when the type is constructed.

Reply:
Ok, will do that once the discussion about this implementation is resolved